migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyMigrator.java

```diff
@@ -0,0 +1,206 @@
+    }
+
+    public void init() {
+        if (taxonomyMapping == null) {
+            readTaxonomyMapping();
+        }
+
+        if (taxonomyDefinition == null) {
+            readTaxonomyDefinition();
+        }
+    }
+
+    public List<String> getTaxonomyKeys(PublishingPackage dataset) {
+        List<String> taxonomyKeys = taxonomyMapping.get(dataset.getUniqueIdentifier());
+        if (taxonomyKeys == null || taxonomyKeys.isEmpty()) {
+            migrationReport.add("No taxonomy terms were provided in the mapping for dataset: " + dataset,
+                "Dataset will be imported without any taxonomy terms.");
+            return Collections.emptyList();
+        }
+
+        Map<Boolean, List<String>> map = taxonomyKeys.stream()
+            .collect(Collectors.partitioningBy(key -> this.taxonomyKeys.contains(key)));
+
+        List<String> notAllowedKeys = map.get(false);
+        if (!notAllowedKeys.isEmpty()) {
+            migrationReport.add("Invalid taxonomy terms were present in the taxonomy mapping",
+                "Dataset: " + dataset,
+                "Taxonomy Terms: " + String.join(", ", notAllowedKeys));
+        }
+
+        return map.get(true);
+    }
+
+    public TaxonomyTerm getTaxonomyDefinition() {
+        return taxonomyDefinition;
+    }
+
+    private void readTaxonomyMapping() {
+        Path taxonomyMappingImportPath = executionParameters.getTaxonomyMappingImportPath();
+        if (taxonomyMappingImportPath == null) {
+            return;
+        }
+
+        Iterator<Row> rowIterator = getRowIterator(taxonomyMappingImportPath);
+
+        Row headerRow = rowIterator.next();
+        List<Integer> pCodeCols = streamRow(headerRow)
+            .filter(cell -> cell.getStringCellValue().equals(P_CODE_COLUMN))
+            .map(Cell::getColumnIndex)
+            .collect(toList());
+
+        if (pCodeCols.size() != 1) {
+            throw new RuntimeException("Spreadsheet needs to have a single P Code column.");
+        }
+
+        int pCodeCol = pCodeCols.get(0);
+
+        List<Integer> taxonomyColumnIndexes = streamRow(headerRow)
+            .filter(cell -> cell.getStringCellValue().startsWith(TAXONOMY_MAPPING_COLUMN_PREFIX))
+            .map(Cell::getColumnIndex)
+            .collect(toList());
+
+        taxonomyMapping = new HashMap<>();
+
+        while (rowIterator.hasNext()) {
+            Row row = rowIterator.next();
+
+            String pCode = row.getCell(pCodeCol).getStringCellValue().trim();
+
+            List<String> taxonomyKeys = streamRow(row)
+                .filter(c -> taxonomyColumnIndexes.contains(c.getColumnIndex()))
+                .map(Cell::getStringCellValue)
+                .map(TaxonomyTerm::covertTermToKey)
+                .filter(key -> !key.isEmpty())
+                .collect(toList());
+
+            if (taxonomyMapping.put(pCode, taxonomyKeys) != null) {
+                throw new RuntimeException("Duplicate taxonomy mappings for pCode: " + pCode);
```

Comment:
As discussed, let's log this in the migration report instead and let the import continue.

Reply:
Will do